packages/flutter_tools/test/general.shard/windows/visual_studio_test.dart

```diff
@@ -143,60 +164,127 @@ void main() {
+      expect(visualStudio.isComplete, false);
     }, overrides: <Type, Generator>{
       FileSystem: () => memoryFilesystem,
       Platform: () => windowsPlatform,
       ProcessManager: () => mockProcessManager,
     });
 
-    testUsingContext('VS metadata is available when VS is present, even if missing components', () {
-      setMockIncompleteVisualStudioExists(true);
-      setMockCompatibleVisualStudioExists(false);
+    testUsingContext('isLaunchable returns false if the installation can\'t be launched', () {
+      setMockCompatibleVisualStudioInstallation(null);
+      setMockPrereleaseVisualStudioInstallation(null);
+
+      final Map<String, dynamic> response = Map<String, dynamic>.from(_defaultResponse)
+        ..['isLaunchable'] = false;
+      setMockAnyVisualStudioInstallation(response);
 
       visualStudio = VisualStudio();
-      expect(visualStudio.displayName, equals('Visual Studio Community 2017'));
-      expect(visualStudio.displayVersion, equals('15.9.12'));
-      expect(visualStudio.installLocation, equals(visualStudioPath));
-      expect(visualStudio.fullVersion, equals('15.9.28307.665'));
+      expect(visualStudio.isInstalled, true);
+      expect(visualStudio.isLaunchable, false);
     }, overrides: <Type, Generator>{
       FileSystem: () => memoryFilesystem,
       Platform: () => windowsPlatform,
       ProcessManager: () => mockProcessManager,
     });
 
+    testUsingContext('isRebootRequired returns true if the installation needs a reboot', () {
+      setMockCompatibleVisualStudioInstallation(null);
+      setMockPrereleaseVisualStudioInstallation(null);
 
-    testUsingContext('isInstalled returns true when VS is present but missing components', () {
-      setMockIncompleteVisualStudioExists(true);
-      setMockCompatibleVisualStudioExists(false);
+      final Map<String, dynamic> response = Map<String, dynamic>.from(_defaultResponse)
+        ..['isRebootRequired'] = true;
+      setMockAnyVisualStudioInstallation(response);
 
       visualStudio = VisualStudio();
       expect(visualStudio.isInstalled, true);
+      expect(visualStudio.isRebootRequired, true);
+    }, overrides: <Type, Generator>{
+      FileSystem: () => memoryFilesystem,
+      Platform: () => windowsPlatform,
+      ProcessManager: () => mockProcessManager,
+    });
+
+
+    testUsingContext('hasNecessaryComponents returns false when VS is present but missing components', () {
+      setMockCompatibleVisualStudioInstallation(null);
+      setMockPrereleaseVisualStudioInstallation(null);
+      setMockAnyVisualStudioInstallation(_defaultResponse);
+
+      visualStudio = VisualStudio();
+      expect(visualStudio.hasNecessaryComponents, false);
+    }, overrides: <Type, Generator>{
+      FileSystem: () => memoryFilesystem,
+      Platform: () => windowsPlatform,
+      ProcessManager: () => mockProcessManager,
+    });
+
+    testUsingContext('vcvarsPath returns null when VS is present but missing components', () {
```

Comment:
I don't know if it's possible for the components to show as present but the install to be unusable due to one of the flags we are checking. To err on the safe side, let's also test some cases where components are present but one of the bad flags (e.g., needs reboot) are true. That will require some code changes to pass I believe—maybe change the logic of how `_usableVisualStudioDetails` is computed internally so that if one of the checks there finds a version that has one of the problem flags set, it's assigned to `_cachedAnyVisualStudioDetails` instead of `_cachedUsableVisualStudioDetails`.

Reply:
Done